text/3735-field-projections.md

```diff
@@ -0,0 +1,1535 @@
+    phantom2: PhantomData<fn(T)>,
+}
+```
+
+### Field Projection Operator
+
+The field projection operator `->` has the following syntax:
+
+> **<sup>Syntax</sup>**\
+> _ProjectionExpression_ :\
+> &nbsp;&nbsp; [_Expression_] `->` _ProjectionMember_
+>
+> _ProjectionMember_ :\
+> &nbsp;&nbsp; &nbsp;&nbsp; [IDENTIFIER]
+> &nbsp;&nbsp; | [TUPLE_INDEX]
+
+[IDENTIFIER]: https://doc.rust-lang.org/reference/identifiers.html
+[_Expression_]: https://doc.rust-lang.org/reference/expressions.html
+[TUPLE_INDEX]: https://doc.rust-lang.org/reference/tokens.html#tuple-index
+
+#### Desugaring
+
+```rust
+struct T {
+    field: F,
+}
+
+let t: C<T> = /* ... */;
+let _ = t->field;
+
+// becomes
+
+let _ = Project::<field_of!(<C<T> as Projectable>::Inner, field)>::project(t);
+```
+
+The `C<T>` in the `C<T> as Projectable` comes from a type inference variable over the expression
+`t`.
+
+
+#### `Projectable` & `Project` Traits
+
+Added to `core::ops`:
+
+```rust
+pub trait Projectable: Sized {
+    type Inner: ?Sized;
+}
+
+pub trait Project<F>: Projectable
+where
+    F: Field<Base = Self::Inner>,
+{
+    type Output;
+
+    fn project(self) -> Self::Output;
+}
+```
+
+## Stdlib Field Projections
+
+All examples from the guide-level explanation work when the standard library is extended with the
+implementations detailed below.
+
+The following pointer types get an implementation for `Projectable` with `Inner = T`. They support
+projections for any field and perform the obvious offset operation.
+
+- `&mut T`
+- `&T`
+- `*mut T`
+- `*const T`
+- `NonNull<T>`
+- `Cow<'_, T>`
+
+For example, `&T` would be implemented like this:
+
+```rust
+impl<'a, T: ?Sized> Projectable for &'a T {
+    type Inner = T;
+}
+
+impl<'a, T: ?Sized, F> Project<F> for &'a T
+where
+    F: Field<Base = T>,
+    // Needed to be able to `.cast` below
+    F::Type: Sized,
+{
+    type Output = &'a F::Type;
+
+    fn project(self) -> Self::Output {
+        let ptr: *const T = self;
+        let ptr = ptr.cast::<u8>();
+        let ptr = unsafe { ptr.add(F::OFFSET) };
+        let ptr = ptr.cast::<F::Type>();
+        unsafe { &*ptr }
+    }
+}
+```
+
+And `Cow` would be implemented like this:
+
+```rust
+impl<'a, T: ?Sized + ToOwned<Owned = T>> Projectable for Cow<'a, T> {
+    type Inner = T;
+}
+
+impl<'a, T: ?Sized + ToOwned<Owned = T>, F> Project<F> for Cow<'a, T>
+where
+    F: Field<Base = T>,
+    F::Type: Sized,
+{
+    type Output = Cow<'a, F::Type>;
+
+    fn project(self) -> Self::Output {
+        match self {
+            Cow::Borrowed(this) => Cow::Borrowed(<&T as Project::<F>>::project(this)),
+            Cow::Owned(this) => Cow::Owned(project_value::<T, F>(this)),
+        }
+    }
+}
+```
+
+Where `project_value` is defined as:
+
+```rust
+fn project_value<T, F>(value: T) -> F::Type
+where
+    F: Field<Base = T>,
+    F::Type: Sized,
+{
+    let r = &value;
+    let r = <&T as Project<F>>::project(r);
+    let res = std::ptr::read(r);
+    std::mem::forget(value);
```

Comment:
this seems highly questionable to me, I'd expect by-value projection to drop all other fields. imo if dropping other fields isn't possible it'd be better to not support projection on `Cow`

Reply:
Oh yeah, that is unintentional... I don't think there is a way to drop all fields except one, or am I missing something?

Reply:
there is a way, but you have to put the field in the source code instead of passing it in with a generic:
```rust
let Struct { field, .. } = {value}; // or just access using value.field
```

if we add `trait MovableField: Field` that has a method to extract the field and drop the rest, then it'll work with generics.

Reply:
That seems possible, but should probably be not part of the rfc, but rather a future possibility. Seems a bit much to make projections for `Cow` work, I personally haven't really needed them there. Just thought of them on a whim. I will move it into future possibilities.